specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,58 @@
 
 ## Overview
 
-// TODO: this is a stub.
+Hooks are a mechanism which can tie into the lifecycle of flag evaluation. They
+operate similarly to middleware in many web frameworks.
+
+### Hook context
+
+Hook context exists to provide hooks with information about the invocation.
+
+> Hook context **MUST** provide: the flag key, flag type, evaluation context, default value and a list of  executed hooks by stage.
+
+> Hook context **SHOULD** provide: provider, client
+
+> flag key, flag type, default value properties **MUST** be immutable.
+
+> The evaluation context **MUST** be mutable only within the `before` hook.
+
+
+### Hook creation and parameters
+
+> Hooks **MUST** specify at least one stage: `before(HookContext, ImmutableState)`, `after(HookContext, FlagEvaluationDetails, , ImmutableState)`, `error(HookContext, Exception, ImmutableState)`, or `finally(HookContext, ImmutableState)`.
```

Comment:
```suggestion
> Hooks **MUST** specify at least one stage: `HookContext | void before(HookContext, ImmutableState)`, ` FlagEvaluationDetails | void after(HookContext, FlagEvaluationDetails, , ImmutableState)`, `void error(HookContext, Exception, ImmutableState)`, or `void finally(HookContext, ImmutableState)`.
```

Reply:
I added the return values (which I think are fairly important).

I wonder if it's better to use plan prose instead of pseudo-code to describe the parameters and return values.

Reply:
> Would be interested to hear what FlagEvaluationDetails return value is used for.

The idea was you could use it for for a few interesting things in after. For example. maybe you want to turn some JSON flag value into a class instance - you could mutate the `.value` field to actually return an instance - the hook would be a little factory that calls a constructor.

Reply:
@toddbaert This would imply that subsequent hooks couldn't assume that the flag value is the type that the flag says it is, right?

Reply:
No, but we might support an generic arg in order to provide that type information in a hook definition. I'm not 100% convinced we need this feature, but it seems like it could be useful.

Reply:
https://github.com/open-feature/spec/issues/48